pbm/pbmtoppa/pbmtoppa.1
.TH pbmtoppa 1 "01 May 2000"
.IX pbmtoppa
.SH NAME
pbmtoppa - convert PBM image to HP Printer Performance Architecture (PPA)
.SH SYNOPSIS
.B pbmtoppa

.RI [ pbm_file 
.RI [ ppa_file ]]

.SH DESCRIPTION
.B pbmtoppa
converts page images in PBM format to Hewlett Packard's PPA (Printer
Performance Architecture) format, which is the data stream format
expected by some HP "Windows-only" printers including the HP Deskjet
820C series, the HP DeskJet 720 series, and the HP DeskJet 1000
series.

.I pbm_file
is the file specification of the input file or 
.B -
for Standard Input.  The default is Standard Input.

The input file contains one or more PBM images, with each one being a
single page.  Each image must have the exact dimensions of a page (at
600 pixels per inch in both directions).  Significantly, this is the
format the Ghostscript produces.

.I ppa_file
is the file specification of the output file or
.B -
for Standard Output.  The default is Standard Output.


To print Postscript on an HP PPA printer, just use Ghostscript with
the
.B pbmraw
(or
.BR pbm )
device driver.

You can generate a test page for use with this program with 
.BR pbmpage .

You can also set up a printer filter so you can submit PBM input
directly to your print queue.  See the documentation for your print
spooler for information on how to do that, or look in hp820install.doc
for an example lpd print filter for Postscript and text files.

Sometimes,
.B pbmtoppa
generates a file which the printer will not print (because 
.BR pbmtoppa 's
input is unprintable).  When this happens, all three lights blink to
signal the error.  This is usually because there is material outside
of the printer's printable area.  To make the file print, increase the
margins via
.B pbmtoppa
options or a configuration file.  See the CALIBRATION section below.

.SH OPTIONS
.TP
.BI "-v " version    
printer version (720, 820, or 1000)
.TP
.BI "-x " xoff
vertical offset adjustment in 1"/600
.TP
.BI "-y " yoff
horizontal offset adjustment in 1"/600
.TP
.BI "-t " topmarg
top margin in 1"/600    (default: 150 = 0.25")
.TP
.BI "-l " leftmarg
left margin in 1"/600   (default: 150 = 0.25")
.TP
.BI "-r " rightmarg
right margin in 1"/600  (default: 150 = 0.25")
.TP
.BI "-b " botmarg
bottom margin in 1"/600 (default: 150 = 0.25")
.TP
.BI "-s " paper
paper size:
.B us  
or
.BR a4 .
Default is 
.BR us .
.TP
.BI "-f " cfgfile
read parameters from configuration file
.I cfgfile 

.PP
The 
.B -x 
and 
.B -y 
options accumulate.  

The 
.B -v 
option resets the horizontal and vertical adjustments to an internal
default.



.SH CONFIGURATION FILES

You can use configuration files to specify parameters rather than use
invocation options.  
.B pbmtoppa
processes the file /etc/pbmtoppa.conf, if it exists, before processing
any options.  It then processes each configuration file named by a 
.B -f
option in order, applying the parameters from the configuration file as
if they were invocation options used in the place of the 
.B -f
option.

Configuration files have the following format:

.BI # Comment
.br
.I key1
.I value1
.br
.I key2
.I value2
.br
[etc.]

Valid 
.IR key s 
are
.BR version ,
.BR xoffset ,
.BR yoffset ,
.BR topmargin ,
.BR leftmargin ,
.BR rightmargin ,
.BR bottommargin ,
.BR papersize ,
or any non-null prefix of these words.  Valid values are the same as
with the corresponding invocation parameters.


.SH EXAMPLES

Print a test pattern: 

.B pbmpage | pbmppa >/dev/lp1

Print three pages:

.B cat page1.pbm page2.pbm page3.pbm | pbmppa >/dev/lp1

Print the Postscript file myfile.ps:

\fBgs -sDEVICE=rawpbm -q -dNOPAUSE -r600 \\
.br
   -sOutputFile=- myfile.ps \\
.br
| pbmtoppa | lpr
\fR

.SH CALIBRATION
To be able to print successfully and properly, you need to tell 
.B pbmtoppa
an X and a Y offset appropriate for your printer to use when
generating the page.  You can specify these offsets with the
.B -x 
and
.B -y
invocation options or with the
.B xoff
and 
.B yoff
parameters in a 
.B pbmtoppa
configuration file.

To determine the correct offsets, use the
.B pbmpage
program.

If while trying to do this calibration, the printer refuses to print a page,
but just blinks all three lights, specify large margins (e.g. 600 pixels --
one inch) via
.B pbmpage
invocation options while doing the calibration.

For example:

.B pbmpage | pbmtoppa >/dev/lp1
.br
or
.br
.B pbmpage | pbmtoppa | lpr -l
.br
(if your printer filter recognizes the '-l' (direct output) parameter).

In the test pattern, the grid is marked off in pixel coordinate numbers.
Unfortunately, these coordinates are probably cut off before the edge of the
paper.  You'll have to use a ruler to estimate the pixel coordinate of the 
left and top edges of the actual sheet of paper (should be within +/- 300, may
be negative; there are 600 pixels per inch).

Add these coordinates to the X and Y offsets by either editing the
configuration file or using the
.B -x 
and 
.B -y
command-line parameters.

When 
.B pbmtoppa
is properly calibrated, the center mark should be in the center of the
paper.  Also, the margins should be able to be as small as 1/4 inch
without causing the printer to choke with 'blinking lights syndrome'.

.SH REDHAT LINUX INSTALLATION

RedHat users may find the following tip from Panayotis Vryonis
<vrypan@hol.gr> helpful.  The same should work for the 820 and 1000,
but it hasn't been tested.  Also, use the pbmraw GSDriver if you have
it; it's faster.

Here is a tip to intergrate HP720C support in RedHat's printtool:

Install pbm2ppa. Copy pbm2ppa to /usr/bin.

Edit "printerdb" (in my system it is found in
/usr/lib/rhs/rhs-printfilters ) and append the following lines:

----------------------Cut here-----------------------
.br
StartEntry: DeskJet720C
  GSDriver: pbm
  Description: {HP DeskJet 720C}
  About: { \\
        This driver supports the HP DeskJet 720C \\
        inkjet printer. \\
        It does does not support color printing. \\
        IMPORTANT! Insert \\
             "- | pbm2ppa -" \\
        in the "Extra GS Otions" field.\\
      }
.br
  Resolution: {600} {600} {}
.br
EndEntry
----------------------------------------------------

Now you can add an HP720C printer just like any other, using printtool.


.SH "SEE ALSO"
.BR pbmpage (1),
.BR pstopnm (1),
.BR pbm (5)

.B pnm2ppa
is not part of Netpbm, but does the same things as 
.B pbmtoppa
except it also works with color and has lots more features.
See <http://sourceforge.net/project/?group_id=1322>.

The file INSTALL-MORE in the pbmtoppa directory of the Netpbm source
code contains detailed instructions on setting up a system to use
pbmtoppa to allow convenient printing on HP PPA printers.  It was
written by Michael Buehlmann.

For information about the PPA protocol and the separately distributed
pbm2ppa program from which
.B pbmtoppa
was derived, see <http://www.httptech.com/ppa>.

.SH AUTHOR
Tim Norman.  Copyright (C) 1998.  Licensed under GNU Public License

Manual page by Bryan Henderson, May 2000.


pbm/pbmtoppa/pbmtoppa.c
/* pbmtoppa.c
 * program to print a 600 dpi PBM file on the HP DJ820Cse or the HP720 series.
 * Copyright (c) 1998 Tim Norman.  See LICENSE for details
 * 2-24-98
 */

#define _BSD_SOURCE
    /* This makes sure strcasecmp() is in string.h */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include "pbm.h"
#include "ppa.h"
#include "ppapbm.h"
#include "cutswath.h"

#include "defaults.h"

/* Paper sizes in 600ths of an inch. */

/* US is 8.5 in by 11 in */

#define USWIDTH  (5100)
#define USHEIGHT (6600)

/* A4 is 210 mm by 297 mm == 8.27 in by 11.69 in */

#define A4WIDTH  (4960)
#define A4HEIGHT (7016)

int Width;      /* width and height in 600ths of an inch */
int Height;
int Pwidth;     /* width in bytes */

#define MAX_LINES 300

ppa_stat printer;

static int 
print_pbm (FILE *in)
{
  char line[1024];
  pbm_stat pbm;
  int done_page, numpages = 0;
  ppa_sweep_data sweeps[2];
  int current_sweep, previous_sweep;

  ppa_init_job(&printer);

  while(make_pbm_stat(&pbm,in))
  {
    if (pbm.width != Width || pbm.height != Height) 
    {
       fprintf(stderr, "print_pbm(): Input image is not the size "
               "of a page for Page %d.\n"
               "The input is %dW x %dH, while a page is %dW x %dH pixels.\n"
               "Page size is controlled by options and the configuration "
               "file.\n",
               numpages+1, pbm.width, pbm.height, Width, Height);
       return 1;
    }

    ppa_init_page(&printer);
    ppa_load_page(&printer);

    sweeps[0].direction = right_to_left;
    sweeps[0].next=&sweeps[1];
    sweeps[1].direction = left_to_right;
    sweeps[1].next=&sweeps[0];

    current_sweep=0;
    previous_sweep=-1;

    done_page=0;
    while(!done_page)
      switch(cut_pbm_swath(&pbm,&printer,MAX_LINES,&sweeps[current_sweep]))
      {
      case 0:
	fprintf (stderr, "print_pbm(): error calling cut_pbm_swath()\n");
	return 1;
      case 1:
	done_page=1;
	break;
      case 2:
	if(previous_sweep>=0)
	{
	  ppa_print_sweep(&printer,&sweeps[previous_sweep]);
	  free(sweeps[previous_sweep].image_data);
	  free(sweeps[previous_sweep].nozzle_data);
	}
	previous_sweep=current_sweep;
	current_sweep= current_sweep==0 ? 1 : 0;
	break;
      default:
	fprintf(stderr,"print_pbm(): unknown return code from cut_pbm_swath()\n");
	return 1;
      }

    if(previous_sweep>=0)
    {
      sweeps[previous_sweep].next=NULL;
      ppa_print_sweep(&printer,&sweeps[previous_sweep]);
    }

    free(sweeps[0].image_data);
    free(sweeps[0].nozzle_data);
    free(sweeps[1].image_data);
    free(sweeps[1].nozzle_data);

    ppa_eject_page(&printer);

    /* eat any remaining whitespace */
    if(pbm.version==P1)
      fgets (line, 1024, in);

    numpages++;
  }

  if (numpages == 0)
  {
    fprintf (stderr, "No pages printed!\n");
    return 1;
  }

  ppa_end_print(&printer);

  fclose (pbm.fptr);
  fclose (printer.fptr);

  return 0;
}



static void 
set_printer_specific_defaults()
{
  switch(printer.version)
  {
  case HP720:
    printer.marg_diff     = HP720_MARG_DIFF;
    printer.bufsize       = HP720_BUFSIZE;

    printer.x_offset      = HP720_X_OFFSET;
    printer.y_offset      = HP720_Y_OFFSET;
    printer.top_margin    = HP720_TOP_MARGIN;
    printer.left_margin   = HP720_LEFT_MARGIN;
    printer.right_margin  = HP720_RIGHT_MARGIN;
    printer.bottom_margin = HP720_BOTTOM_MARGIN;
    break;
  case HP820:
    printer.marg_diff     = HP820_MARG_DIFF;
    printer.bufsize       = HP820_BUFSIZE;

    printer.x_offset      = HP820_X_OFFSET;
    printer.y_offset      = HP820_Y_OFFSET;
    printer.top_margin    = HP820_TOP_MARGIN;
    printer.left_margin   = HP820_LEFT_MARGIN;
    printer.right_margin  = HP820_RIGHT_MARGIN;
    printer.bottom_margin = HP820_BOTTOM_MARGIN;
    break;
  case HP1000:
    printer.marg_diff     = HP1000_MARG_DIFF;
    printer.bufsize       = HP1000_BUFSIZE;

    printer.x_offset      = HP1000_X_OFFSET;
    printer.y_offset      = HP1000_Y_OFFSET;
    printer.top_margin    = HP1000_TOP_MARGIN;
    printer.left_margin   = HP1000_LEFT_MARGIN;
    printer.right_margin  = HP1000_RIGHT_MARGIN;
    printer.bottom_margin = HP1000_BOTTOM_MARGIN;
    break;
  default:
    fprintf(stderr,"set_printer_defaults(): unknown printer version\n");
    exit(1);
  }
}



static void 
show_usage(char* prog)
{
  printf("usage: %s [ options ] [ <infile> [ <outfile> ] ]\n\n",prog);
  printf("  Prints a pbm- or pbmraw-format <infile> to HP720/820/1000-format <outfile>.\n\n");
  printf("    -v <version>    printer version (720, 820, or 1000)\n");
  printf("    -x <xoff>       vertical offset adjustment in 1\"/600\n");
  printf("    -y <yoff>       horizontal offset adjustment in 1\"/600\n");
  printf("    -t <topmarg>    top margin in 1\"/600    (default: 150 = 0.25\")\n");
  printf("    -l <leftmarg>   left margin in 1\"/600   (default: 150 = 0.25\")\n");
  printf("    -r <rightmarg>  right margin in 1\"/600  (default: 150 = 0.25\")\n");
  printf("    -b <botmarg>    bottom margin in 1\"/600 (default: 150 = 0.25\")\n");
  printf("    -s <paper>      paper size (us, a4, default: us)\n");
  printf("    -f <cfgfile>    read <cfgfile> as parameters\n\n");
  printf("  The -x and -y options accumulate.  The -v option resets the horizontal and\n");
  printf("  vertical adjustments to an internal default.  <infile> and <outfile> default\n");
  printf("  to stdin and stdout.  '-' is a synonym for stdin and stdout.\n\n");
  printf("  Configuration files specified with the '-f' parameter have the following\n  format:\n\n");
  printf("    # Comment\n");
  printf("    <key1> <value1>\n");
  printf("    <key2> <value2>\n");
  printf("    [etc.]\n\n");
  printf("  Valid keys are 'version', 'xoffset', 'yoffset', 'topmargin', 'leftmargin',\n");
  printf("  'rightmargin', 'bottommargin', 'papersize', or any non-null truncated\n");
  printf("  version of these words.  Valid values are the same as with the corresponding\n");
  printf("  command-line parameters.  Parameters in the configuration file act as though\n");
  printf("  the corresponding parameters were substituted, in order, for the '-f'\n");
  printf("  parameter which specified the file.\n\n");
  printf("  The file /etc/pbmtoppa.conf, if it exists, is processed as a configuration\n");
  printf("  file before any command-line parameters are processed.\n\n");
}



static void 
parm_version(char* arg)
{
  if(!strcasecmp(arg,"hp720") || !strcmp(arg,"720"))
    printer.version=HP720;
  else if(!strcasecmp(arg,"hp820") || !strcmp(arg,"820"))
    printer.version=HP820;
  else if(!strcasecmp(arg,"hp1000") || !strcmp(arg,"1000"))
    printer.version=HP1000;
  else
  {
    fprintf(stderr,"parm_version(): unknown printer version '%s'\n",arg);
    exit(1);
  }
  set_printer_specific_defaults();
}



static void 
parm_iversion(int arg)
{
  switch(arg)
  {
  case 720:
    printer.version=HP720;
    break;
  case 820:
    printer.version=HP820;
    break;
  case 1000:
    printer.version=HP1000;
    break;
  default:
    {
      fprintf(stderr,"parm_iversion(): unknown printer version '%d'\n",arg);
      exit(1);
    }
  }
  set_printer_specific_defaults();
}



static void 
dump_config()
{
  printf("version:  ");
  switch(printer.version)
  {
  case HP710:  printf("HP710\n");  break;
  case HP720:  printf("HP720\n");  break;
  case HP820:  printf("HP820\n");  break;
  case HP1000: printf("HP1000\n"); break;
  }
  printf("x-offset: %d\ny-offset: %d\nmargins:\n top:    %d\n"
	 " left:   %d\n right:  %d\n bottom: %d\n",printer.x_offset,
	 printer.y_offset,printer.top_margin,printer.left_margin,
	 printer.right_margin,printer.bottom_margin);
  exit(0);
}



static void 
read_config_file(char* fname)
{
  FILE* cfgfile=fopen(fname,"r");
  char line[1024],key[14],buf[10];
  int len,value,lineno=1;

  if(!cfgfile)
  {
    fprintf (stderr, "read_config_file(): couldn't open file '%s'\n", fname);
    exit(1);
  }

  while(fgets(line,1024,cfgfile))
  {
    if(strchr(line,'#'))
      *strchr(line,'#')=0;
    switch(sscanf(line,"%13s%9s",key,buf))
    {
    case 2:
      value=atoi(buf);
      len=strlen(key);
      if(!strncmp(key,"version",len))
	parm_iversion(value);
      else if(!strncmp(key,"xoffset",len))
	printer.x_offset=value;
      else if(!strncmp(key,"yoffset",len))
	printer.y_offset=value;
      else if(!strncmp(key,"topmargin",len))
	printer.top_margin=value;
      else if(!strncmp(key,"leftmargin",len))
	printer.left_margin=value;
      else if(!strncmp(key,"rightmargin",len))
	printer.right_margin=value;
      else if(!strncmp(key,"bottommargin",len))
	printer.bottom_margin=value;
      else if(!strncmp(key,"papersize",len))
      {
	if(!strcmp(buf,"us"))
	{
	  Width = USWIDTH;
	  Height = USHEIGHT;
	}
	else if(!strcmp(buf,"a4"))
	{
	  Width = A4WIDTH;
	  Height = A4HEIGHT;
	}
	else
	{
	  fprintf(stderr,"read_config_file(): unknown paper size %s\n",buf);
	  exit(1);
	}
      }
      else if(!strcmp(key,"dump"))
	dump_config();
      else 
      {
	fprintf(stderr,"read_config_file(): unrecognized parameter '%s' (line %d)\n\n",key,lineno);
	show_usage("pbmtoppa");
	exit(1);
      }
    case EOF:
    case 0: break;
    default:
      fprintf(stderr,"read_config_file(): error parsing config file (line %d)\n",lineno);
      exit(1);
    }
    lineno++;
  }

  if(feof(cfgfile))
  {
    fclose(cfgfile);
    return;
  }

  fprintf(stderr,"read_config_file(): error parsing config file\n");
  exit(1);
}



char* defaultcfgfile="/etc/pbmtoppa.conf";



int 
main (int argc, char *argv[])
{
  int argn;
  int got_in=0, got_out=0, do_continue=1;
  FILE *in=stdin, *out=stdout;
  struct stat tmpstat;

  pbm_init(&argc, argv);

  printer.version       = DEFAULT_PRINTER;
  printer.x_offset      = DEFAULT_X_OFFSET;
  printer.y_offset      = DEFAULT_Y_OFFSET;
  printer.top_margin    = DEFAULT_TOP_MARGIN;
  printer.left_margin   = DEFAULT_LEFT_MARGIN;
  printer.right_margin  = DEFAULT_RIGHT_MARGIN;
  printer.bottom_margin = DEFAULT_BOTTOM_MARGIN;
  Width = USWIDTH;
  Height = USHEIGHT;
  set_printer_specific_defaults();

  if(!stat(defaultcfgfile,&tmpstat))
    read_config_file(defaultcfgfile);

  for(argn=1; argn<argc; argn++)
  {
    if(!strcmp(argv[argn],"-h"))
    {
      show_usage(*argv);
      return 0;
    }
    else if(!strcmp(argv[argn],"-d"))
      dump_config();
    else if(argn+1<argc)
    {
      do_continue=1;
      if(!strcmp(argv[argn],"-v"))
	parm_version(argv[++argn]);
      else if(!strcmp(argv[argn],"-x"))
	printer.x_offset+=atoi(argv[++argn]);
      else if(!strcmp(argv[argn],"-y"))
	printer.y_offset+=atoi(argv[++argn]);
      else if(!strcmp(argv[argn],"-t"))
	printer.top_margin=atoi(argv[++argn]);
      else if(!strcmp(argv[argn],"-l"))
	printer.left_margin=atoi(argv[++argn]);
      else if(!strcmp(argv[argn],"-r"))
	printer.right_margin=atoi(argv[++argn]);
      else if(!strcmp(argv[argn],"-b"))
	printer.bottom_margin=atoi(argv[++argn]);
      else if(!strcmp(argv[argn],"-s"))
      {
	argn++;
	if(!strcmp(argv[argn],"us"))
	{
	  Width = USWIDTH;
	  Height = USHEIGHT;
	}
	else if(!strcmp(argv[argn],"a4"))
	{
	  Width = A4WIDTH;
	  Height = A4HEIGHT;
	}
	else
	{
	  fprintf(stderr,"unknown paper size %s\n",argv[argn]);
	  return 1;
	}
      }
      else if(!strcmp(argv[argn],"-f"))
	read_config_file(argv[++argn]);
      else do_continue=0;
      if(do_continue) continue;
    }

    if(!got_in)
    {
      if (strcmp (argv[argn], "-") == 0)
	in = stdin;
      else if ((in = fopen (argv[argn], "rb")) == NULL)
      {
	fprintf (stderr, "main(): couldn't open file '%s'\n", argv[argn]);
	return 1;
      }
      got_in=1;
    }
    else if(!got_out)
    {
      if (strcmp (argv[argn], "-") == 0)
	out = stdout;
      else if ((out = fopen (argv[argn], "wb")) == NULL)
      {
	fprintf (stderr, "main(): couldn't open file '%s'\n", argv[argn]);
	return 1;
      }
      got_out=1;
    }
    else
    {
      fprintf(stderr,"main(): unrecognized parameter '%s'\n",argv[argn]);
      show_usage(*argv);
      return 1;
    }
  }

  overflow_add(Width, 7);
  Pwidth=(Width+7)/8;
  printer.fptr=out;

  return print_pbm (in);
}

